File: app/routers/qa.py
"""
GET /api/v1/qa       — 分頁列表（支援篩選）
GET /api/v1/qa/{id}  — 取單筆
GET /api/v1/qa/categories — 所有分類
"""
from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, HTTPException, Query, Request
from pydantic import BaseModel

from app.core.store import QAItem, store
from utils import audit_logger

router = APIRouter(prefix="/api/v1/qa", tags=["qa"])


# ──────────── schema ────────────

class QAResponse(BaseModel):
    id: int
    question: str
    answer: str
    keywords: list[str]
    confidence: float
    category: str
    difficulty: str
    evergreen: bool
    source_title: str
    source_date: str
    is_merged: bool


class QAListResponse(BaseModel):
    items: list[QAResponse]
    total: int
    offset: int
    limit: int


class CategoriesResponse(BaseModel):
    categories: list[str]


def _to_schema(item: QAItem) -> QAResponse:
    return QAResponse(
        id=item.id,
        question=item.question,
        answer=item.answer,
        keywords=item.keywords,
        confidence=item.confidence,
        category=item.category,
        difficulty=item.difficulty,
        evergreen=item.evergreen,
        source_title=item.source_title,
        source_date=item.source_date,
        is_merged=item.is_merged,
    )


# ──────────── routes ────────────

@router.get("/categories", response_model=CategoriesResponse)
def get_categories() -> CategoriesResponse:
    return CategoriesResponse(categories=store.categories())


@router.get("/{item_id}", response_model=QAResponse)
def get_qa_item(item_id: int) -> QAResponse:
    for item in store.items:
        if item.id == item_id:
            return _to_schema(item)
    raise HTTPException(status_code=404, detail=f"QA id={item_id} not found")


@router.get("", response_model=QAListResponse)
def list_qa(
    request: Request,
    category: Optional[str] = Query(default=None),
    keyword: Optional[str] = Query(default=None, max_length=100),
    difficulty: Optional[str] = Query(default=None, pattern="^(基礎|進階)$"),
    evergreen: Optional[bool] = Query(default=None),
    limit: int = Query(default=20, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
) -> QAListResponse:
    items, total = store.list_qa(
        category=category,
        keyword=keyword,
        difficulty=difficulty,
        evergreen=evergreen,
        limit=limit,
        offset=offset,
    )

    # 稽核日誌：記錄列表查詢與返回的 QA IDs
    client_ip = request.client.host if request.client else None
    audit_logger.log_list_qa(
        filters={
            "category": category,
            "keyword": keyword,
            "difficulty": difficulty,
            "evergreen": evergreen,
            "limit": limit,
            "offset": offset,
        },
        returned_ids=[i.id for i in items],
        total=total,
        client_ip=client_ip,
    )

    return QAListResponse(
        items=[_to_schema(i) for i in items],
        total=total,
        offset=offset,
        limit=limit,
    )

File: api/src/routes/qa.ts
import { Hono } from "hono";
import { qaIdPattern, qaListParamsSchema } from "../schemas/qa.js";
import { ok, fail } from "../schemas/api-response.js";
import { qaStore, type QAItem } from "../store/qa-store.js";

export const qaRoute = new Hono();

function toSchema(item: QAItem) {
  return {
    id: item.id,
    seq: item.seq,
    question: item.question,
    answer: item.answer,
    keywords: item.keywords,
    confidence: item.confidence,
    category: item.category,
    difficulty: item.difficulty,
    evergreen: item.evergreen,
    source_title: item.source_title,
    source_date: item.source_date,
    is_merged: item.is_merged,
    notion_url: item.notion_url,
    source_type: item.source_type,
    source_collection: item.source_collection,
    source_url: item.source_url,
    extraction_model: item.extraction_model ?? null,
    maturity_relevance: item.maturity_relevance ?? null,
    primary_category: item.primary_category ?? item.category,
    categories: item.categories ?? [item.category],
    intent_labels: item.intent_labels ?? [],
    scenario_tags: item.scenario_tags ?? [],
    serving_tier: item.serving_tier ?? "canonical",
    evidence_scope: item.evidence_scope ?? [],
    freshness_score: item.freshness_score,
  };
}

qaRoute.get("/categories", (c) => {
  return c.json(ok({ categories: qaStore.categories() }));
});

qaRoute.get("/collections", (c) => {
  return c.json(ok({ collections: qaStore.collections() }));
});

qaRoute.get("/:item_id", (c) => {
  const itemId = c.req.param("item_id");

  // Support both hex stable_id and integer seq
  let item;
  const seqNum = Number(itemId);
  if (Number.isInteger(seqNum) && seqNum > 0) {
    item = qaStore.getBySeq(seqNum);
  } else if (qaIdPattern.test(itemId)) {
    item = qaStore.getById(itemId);
  } else {
    return c.json(fail("Invalid QA ID format"), 400);
  }

  if (!item) {
    return c.json(fail(`QA id=${itemId} not found`), 404);
  }
  return c.json(ok(toSchema(item)));
});

qaRoute.get("/", (c) => {
  const parsed = qaListParamsSchema.safeParse({
    category: c.req.query("category"),
    primary_category: c.req.query("primary_category"),
    keyword: c.req.query("keyword"),
    difficulty: c.req.query("difficulty"),
    evergreen: c.req.query("evergreen"),
    source_type: c.req.query("source_type"),
    source_collection: c.req.query("source_collection"),
    extraction_model: c.req.query("extraction_model"),
    maturity_relevance: c.req.query("maturity_relevance"),
    intent_label: c.req.query("intent_label"),
    scenario_tag: c.req.query("scenario_tag"),
    serving_tier: c.req.query("serving_tier"),
    evidence_scope: c.req.query("evidence_scope"),
    sort_by: c.req.query("sort_by"),
    sort_order: c.req.query("sort_order"),
    limit: c.req.query("limit"),
    offset: c.req.query("offset"),
  });

  if (!parsed.success) {
    return c.json(fail("Invalid query parameters"), 400);
  }

  const { limit, offset, ...filters } = parsed.data;
  const { items, total } = qaStore.listQa({ ...filters, limit, offset });

  return c.json(
    ok({
      items: items.map(toSchema),
      total,
      offset,
      limit,
    }),
  );
});
